dpnp/linalg/dpnp_utils_linalg.py

```diff
@@ -383,6 +385,78 @@ def _lu_factor(a, res_type):
         return (a_h, ipiv_h, dev_info_array)
 
 
+def _multi_dot(arrays, order, i, j, out=None):
+    """Actually do the multiplication with the given order."""
+    if i == j:
+        # the initial call with non-None out should never get here
+        assert out is None
+        return arrays[i]
+
+    return dpnp.dot(
+        _multi_dot(arrays, order, i, order[i, j]),
+        _multi_dot(arrays, order, order[i, j] + 1, j),
+        out=out,
+    )
+
+
+def _multi_dot_matrix_chain_order(n, arrays, return_costs=False):
+    """
+    Return a dpnp.ndarray that encodes the optimal order of mutiplications.
+
+    The optimal order array is then used by `_multi_dot()` to do the
+    multiplication.
+
+    Also return the cost matrix if `return_costs` is ``True``.
+
+    The implementation CLOSELY follows Cormen, "Introduction to Algorithms",
+    Chapter 15.2, p. 370-378.  Note that Cormen uses 1-based indices.
+
+        cost[i, j] = min([
+            cost[prefix] + cost[suffix] + cost_mult(prefix, suffix)
+            for k in range(i, j)])
+
+    """
+
+    # p stores the dimensions of the matrices
+    # Example for p: A_{10x100}, B_{100x5}, C_{5x50} --> p = [10, 100, 5, 50]
+    p = [1 if arrays[0].ndim == 0 else a.shape[0] for a in arrays]
+    p.append(1 if arrays[-1].ndim == 0 else arrays[-1].shape[1])
```

Comment:
Numpy has a bit different behavior. To align you need to have something, like:
```suggestion
    p = [1 if arrays[0].ndim == 1 else arrays[0].shape[0]]
    p += [a.shape[0] for a in arrays[1:-1]]
    p += [1, arrays[-1].shape[0]] if arrays[-1].ndim == 1 else [arrays[-1].shape[0], arrays[-1].shape[1]]
```
otherwise:
```python
a = numpy.ones((10, 100), dtype='f4')
b = numpy.ones((100, 5), dtype='f4')
c = numpy.ones((5), dtype='f4')
c2 = numpy.atleast_2d(c)

from numpy.linalg.linalg import _multi_dot_matrix_chain_order
_multi_dot_matrix_chain_order([a, b, c2], True)
# (array([[0, 0, 1],
#         [0, 0, 1],
#         [0, 0, 0]]),
#  array([[   0., 1000., 1050.],
#         [   0.,    0.,  500.],
#         [   0.,    0.,    0.]]))

ia = dpnp.asarray(a)
ib = dpnp.asarray(b)
ic = dpnp.asarray(c)

from dpnp.linalg.dpnp_utils_linalg import _multi_dot_matrix_chain_order
_multi_dot_matrix_chain_order(3, [ia, ib, ic], True)
# (array([[0, 0, 0],
#         [0, 0, 1],
#         [0, 0, 0]]),
#  array([[   0., 5000., 1500.],
#         [   0.,    0.,  500.],
#         [   0.,    0.,    0.]], dtype=float32))
```
Or was it intended?

Also, probably it would be much more easy just to convert 1d arrays into 2d as numpy did through `atleast_2d` call.

Reply:
your suggested changes are valid and I have updated the code.

However, I am not converting 1d arrays into 2d since it seems there is a bug in NumPy. For instance, this case does not work while out shape is correct.

```
>>> import dpctl,dpctl.tensor as dpt, numpy, dpnp
>>> a = numpy.ones((4,), dtype='f4')
>>> b = numpy.ones((4, 6), dtype='f4')
>>> c = numpy.ones((6, 8), dtype='f4')
>>> res = numpy.empty((8,), order='C', dtype='f4')
>>> numpy.linalg.multi_dot([a,b,c]).shape
(8,)

>>> numpy.linalg.multi_dot([a,b,c], out=res).shape
ValueError: output array is not acceptable (must have the right datatype, number of dimensions, and be a C-Array)
```

The reason is when NumPy changes the shape if `a` in this example to `(1,4)`, it needs the out shape to be `(1, 8)` for intermediate calculations but it is `(8,)`. To avoid this problem in dpnp, I chose to not make 1d arrays 2d.

Reply:
Yep, looks reasonable. Thank you for clarification.